js/passwordManager/passwordManager.js

```diff
@@ -0,0 +1,220 @@
+      console.error('Error accessing Bitwarden CLI. STDOUT: ' + data + '. STDERR: ' + error)
+      return []
+    }
+  }
+
+  // Tries to unlock the store by asking for a master password and
+  // then passing that to Bitwarden-CLI to get a session key.
+  async tryToUnlock(command) {
+    let password = await this.promptForMasterPassword()
+    let sessionKey = await this.unlockStore(command, password)
+    this.sessionKey = sessionKey
+  }
+  
+  // Tries to unlock the password store with given master password. 
+  async unlockStore(command, password) {
+    try {
+      let process = new ProcessSpawner(command, ['unlock', '--raw', password])
+      let result = await process.execute()
+      return result
+    } catch (ex) {
+      const { error, data } = ex
+      console.error('Error accessing Bitwarden CLI. STDOUT: ' + data + '. STDERR: ' + error)
+      throw ex
+    }
+  }
+
+  // Shows a prompt dialog for password store's master password.
+  async promptForMasterPassword() {
+    return new Promise((resolve, reject) => {
+      let { data: password } = ipc.sendSync('prompt', { text: 'Please enter Bitwarden master password to unlock the password store:' })
+      if (password == null || password == '') {
+        reject()
+      } else {
+        resolve(password)
+      }
+    })
+  }
+}
+
+// List of supported password managers. Each password manager is expected to 
+// have getSuggestions(domain) method that returns a Promise with credentials
+// suggestions matching given domain name.
+var passwordManagers = [
+    new Bitwarden()
+]
+
+// Returns an active password manager, which is the one that is selected in app's
+// settings. 
+async function getActivePasswordManager() {
+  return new Promise((resolve, reject) => {
+    if (passwordManagers.length == 0) {
+      resolve(null)
+    }
+
+    let managerSetting = settings.get('passwordManager')
+    if (managerSetting == null) {
+      resolve(null)
+    }
+
+    let manager = passwordManagers.find(mgr => mgr.name == managerSetting.name)
+    resolve(manager)
+  })
+}
+
+async function getConfiguredPasswordManager() {
+  let manager = await getActivePasswordManager()
+  if (!manager) {
+    return null
+  }
+
+  let configured = await manager.checkIfConfigured()
+  if (!configured) {
+    return null
+  }
+
+  return manager
+}
+
+// Called when page preload script detects a form with username and password.
+webviews.bindIPC('password-autofill', function (webview, tab, args, frameId) {
+  getConfiguredPasswordManager().then((manager) => {
+    if (!manager) {
+      return
+    }
+
+    webviews.callAsync(tab, 'getURL', null, (err, src) => {
+      var domain = new URL(src).hostname
+      if (domain.startsWith('www.')) {
+        domain = domain.slice(4)
+      }
+      
+      var self = this
+      manager.getSuggestions(domain).then(credentials => {
+        if (credentials.length > 0) {
+          webview.sendToFrame(frameId, 'password-autofill-match', credentials)
```

Comment:
I'm pretty sure this is a security issue - if the frame that sent the `password-autofill` event is from a third party, then the frame will receive credentials from the first-party site. We need to send the URL of the frame to the password manager, not the URL of the tab.

Reply:
Re-implemented by sending `document.location` as a parameter to this IPC call. As I understand it, frames have their own `documents`, so that should send frame's actual location/hostname.